// create-fp-ts-lib/dist/features/vscode/tasks-json.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var A = __importStar(require("fp-ts/Array"));
var function_1 = require("fp-ts/lib/function");
var task = function (config) { return function (name) { return ({
    label: name + ":watch",
    runOptions: { runOn: 'folderOpen' },
    type: 'shell',
    command: config.packageManager,
    isBackground: true,
    args: ['run', name + ":watch"],
    group: name === 'build' ? 'build' : 'none',
    presentation: {
        reveal: 'always',
        echo: false,
        focus: false,
        panel: 'dedicated',
    },
}); }; };
var tasks = function (config, names) { return ({
    version: '2.0.0',
    tasks: function_1.pipe(names, A.map(task(config))),
}); };
exports.default = tasks;
