// create-fp-ts-lib/dist/features/cspell.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
Object.defineProperty(exports, "__esModule", { value: true });
var RTE = __importStar(require("../ReaderTaskEither"));
var function_1 = require("fp-ts/lib/function");
var type_utils_1 = require("../type-utils");
var Apply_1 = require("fp-ts/lib/Apply");
var PJ = __importStar(require("../PackageJson"));
// -----------------------------------------------------------------------------
// utils
// -----------------------------------------------------------------------------
var mkPackageJson = {
    devDependencies: {
        cspell: '^5.5.2',
    },
    scripts: {
        spell: "yarn cspell '**/*.*'",
    },
};
var mkCspellJson = function (config) { return ({
    version: '0.1',
    language: 'en',
    ignorePaths: __spreadArray(__spreadArray(__spreadArray([
        '**/node_modules/**',
        '**/.cache/**',
        '**/dist/**',
        '**/.git/**',
        '**/.cache/**',
        '**/*.tsbuildinfo',
        'tmp/'
    ], (config.packageManager === 'yarn'
        ? ['yarn.lock', '**/yarn-error.log']
        : ['package-lock.json'])), (config.docsTs ? ['**/docs/examples/', '**/docs/modules/*.ts.md'] : [])), (config.vscode ? ['.vscode'] : [])),
    words: [],
}); };
// -----------------------------------------------------------------------------
// effect
// -----------------------------------------------------------------------------
var packageJson = RTE.scope(function (_a) {
    var data = _a.files["package.json"].data;
    return function_1.pipe(mkPackageJson, PJ.merge(data), type_utils_1.tag('PackageJson'), RTE.of);
});
var cspellJson = RTE.scope(function (_a) {
    var config = _a.config;
    return function_1.pipe(mkCspellJson(config), type_utils_1.tag('Json'), RTE.of);
});
var main = function_1.pipe({
    'package.json': packageJson,
    '.cspell.json': cspellJson,
}, Apply_1.sequenceS(RTE.ApplySeq));
// -----------------------------------------------------------------------------
// export
// -----------------------------------------------------------------------------
exports.default = main;
