// create-fp-ts-lib/dist/features/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cspell = exports.markdownMagic = exports.vscode = exports.ghActions = exports.fastCheck = exports.docsTs = exports.jest = exports.esLint = exports.prettier = exports.skeleton = void 0;
var skeleton_1 = __importDefault(require("./skeleton"));
exports.skeleton = skeleton_1.default;
var prettier_1 = __importDefault(require("./prettier"));
exports.prettier = prettier_1.default;
var eslint_1 = __importDefault(require("./eslint"));
exports.esLint = eslint_1.default;
var jest_1 = __importDefault(require("./jest"));
exports.jest = jest_1.default;
var docs_ts_1 = __importDefault(require("./docs-ts"));
exports.docsTs = docs_ts_1.default;
var fast_check_1 = __importDefault(require("./fast-check"));
exports.fastCheck = fast_check_1.default;
var gh_actions_1 = __importDefault(require("./gh-actions"));
exports.ghActions = gh_actions_1.default;
var vscode_1 = __importDefault(require("./vscode"));
exports.vscode = vscode_1.default;
var markdown_magic_1 = __importDefault(require("./markdown-magic"));
exports.markdownMagic = markdown_magic_1.default;
var cspell_1 = __importDefault(require("./cspell"));
exports.cspell = cspell_1.default;
